flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/async/AsyncCollectorBufferTest.java

```diff
@@ -0,0 +1,293 @@
+
+		AsyncCollectorBuffer<Integer, Integer> buffer;
+		ExecutorService service = Executors.newFixedThreadPool(10);
+
+		boolean throwExcept = false;
+		boolean orderedMode = false;
+
+		public OrderedPutThread(AsyncCollectorBuffer buffer, boolean except, boolean orderedMode) {
+			this.buffer = buffer;
+			this.throwExcept = except;
+			this.orderedMode = orderedMode;
+		}
+
+		public OrderedPutThread(AsyncCollectorBuffer buffer, boolean except) {
+			this(buffer, except, true);
+		}
+
+		public OrderedPutThread(AsyncCollectorBuffer buffer) {
+			this(buffer, false);
+		}
+
+		@Override
+		public void run() {
+			try {
+				for (int idx = 0; idx < ASYNC_COLLECTOR_NUM; ++idx) {
+					int i = orderedSeq[idx];
+					final int sleepTS = sleepTimeArr[idx]*1000;
+
+					if (i == 3)
+						buffer.add(new Watermark(333l));
+					else if (i == 6)
+						buffer.add(new LatencyMarker(111L, 0, 0));
+					else {
+						StreamRecord record = new StreamRecord(i);
+						record.setTimestamp(i*i);
+						final AsyncCollector collector = buffer.add(record);
+
+						final int v = i;
+						service.submit(new Runnable() {
+							@Override
+							public void run() {
+								try {
+									int sleep = sleepTS;
+									Thread.sleep(sleep);
+									if (throwExcept)
+										collector.collect(new Exception("wahahaha..."));
+									else {
+										List<Integer> ret = new ArrayList<Integer>();
+										ret.add(v);
+										collector.collect(ret);
+									}
+								} catch (Exception e) {
+									e.printStackTrace();
+								}
+							}
+						});
+					}
+				}
+			}
+			catch (Exception e) {
+				e.printStackTrace();
+			}
+		}
+	}
+
+	public class FakedOutput implements Output<StreamRecord<Integer>> {
+		private Collection<Long> outputs;
+		private StringBuilder sb = new StringBuilder();
+
+		public FakedOutput(Collection<Long> outputs) {
+			this.outputs = outputs;
+		}
+
+		@Override
+		public void collect(StreamRecord<Integer> record) {
+			outputs.add(record.getValue().longValue());
+			if (record.hasTimestamp())
+				sb.append(record.getTimestamp()+",");
+		}
+
+		@Override
+		public void emitWatermark(Watermark mark) {
+			outputs.add(mark.getTimestamp());
+		}
+
+		@Override
+		public void emitLatencyMarker(LatencyMarker latencyMarker) {
+			outputs.add(latencyMarker.getMarkedTime());
+		}
+
+		@Override
+		public void close() {
+		}
+
+		public String getResult() {
+			StringBuilder sb = new StringBuilder();
+			for (Long i : outputs)
+				sb.append(i).append(",");
+			return sb.toString();
+		}
+
+		public String getTimestamp() {
+			return this.sb.toString();
+		}
+	}
+
+	@Test
+	public void testOrderedBuffer() throws Exception {
+		buffer.startEmitterThread();
+		Watermark watermark = new Watermark(0L);
+		buffer.add(watermark);
+
+		OrderedPutThread orderedPutThread = new OrderedPutThread(buffer);
+		Thread.sleep(2000);
+
+		orderedPutThread.start();
+
+		orderedPutThread.join();
```

Comment:
We have to wait all inputs have been sent out by `PutterThread`, otherwise, the buffer may have a chance to go through `waitEmpty()` and the test may fail.